Add lazy:: backend interfaces

wconstab opened the pull request:
Summary:
BackendData - represents 'tensor data' in opaque backend storage
LoweringContext - interface for performing backend-specific IR lowering
BackendImplInterface - interface for lazy tensors backends to implement

Reorgs backend-related files into lazy/backend subdir

includes a few small fixes, which were made on lazy_tensor_staging but need to be back-ported to master.

Test Plan: used by lazy_tensor_staging branch

Differential Revision: D32142032

JackCaoG commented on torch/csrc/lazy/backend/backend_data.h:
```diff
@@ -0,0 +1,61 @@
+    return device_;
+  }
+
+  const Shape& shape() const {
```
One somewhat related comment. Since LTC decide to keep the `torch::lazy::shape` on the core level, is it a good idea to actually add `torch::lazy::shape` to the Base IR node? This way we don't need to keep the current `getShape(IR)` call. I can also make `XlaNode` to have a `xlaShape()` method and slowly replaced the usage of it to `shape()`.

wconstab replied:
Yea, I think we can add it back to Node base.  Jiewen and you have both been twisting my arm.  :)

I think @Krovatkin had some reservations about it all along; that we don't want shape in the IR simply because we want the IR reusable.  Nick, are you good with this, maybe if we put a comment explaining the shape is useful for upper-bounds inference but shouldn't be relied on in TS backend?  Any other concerns?

Krovatkin replied:
@wconstab 
I think it does make a lot of sense to provide this utility for other vendors, too. 

> maybe if we put a comment explaining the shape is useful for upper-bounds inference but shouldn't be relied on in TS backend? Any other concerns?

I wonder if we could code-document it by maybe renaming `shape` to `upper_bounds` or something like that? Or would it break too many things downstream and in LTC? @JackCaoG

JackCaoG replied:
it doesn't matter too much for pt/xla right now as we will keep using `xlashape()` in the short term. The plan is slowly migrate to use `lazy::shape`  from lazy tensor core. I actually think it is better to name it `shape` since in the static mode `upper_bound` is confusing. Plus `Shape` also encasualte the `c10::ScalarType`.

JackCaoG reviewed (approved):
Thanks!

alanwaketan commented on torch/csrc/lazy/backend/backend_interface.h:
```diff
@@ -0,0 +1,129 @@
+  // scenes. In the future, non-virtual c10:: devices may also use lazy tensors
+  // through a mode, in which case these APIs should still work, but should be
+  // identity mappings.
+  virtual BackendDevice GetBackendDevice(c10::Device device) const = 0;
```
I think we also need the other way round?
```
virtual c10::Device GetCoreDevice(const BackendDevice& device) const = 0;
```

wconstab replied:
what for? (just for symmetry sake, or do you have a use case in mind?) 

Note that I added another API for EagerFallback device.  This API might be what you're looking for. 

Why? 
1) it's probably not that useful to translate from a virtual device like TPU back to the corresponding c10 'XLA' device, since the XLA backend can assume it's using the XLA device.  
2) for places in the code where we need to work with a _real_ c10 device like CPU/cuda (for eager fallback, for example), we don't actually want to be told that TPU maps to XLA, we want to be told that TPU maps to c10::CPU, for fallback, however, GPU should map to c10::CUDA, for more efficient fallback.

desertfire commented on torch/csrc/lazy/backend/backend_data.h:
```diff
@@ -0,0 +1,61 @@
+  virtual bool HasValue() const = 0;
+
+ private:
+  std::string device_;
```
Jiewen has recently updated this to Device. We should use the latest version.

wconstab replied:
oh yea, good catch.  i meant to change that.

wconstab committed:
[LT] Upstream backend interfaces (#67927)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/67927

BackendData - represents 'tensor data' in opaque backend storage
LoweringContext - interface for performing backend-specific IR lowering
BackendImplInterface - interface for lazy tensors backends to implement

Reorgs backend-related files into lazy/backend subdir

includes a few small fixes, which were made on lazy_tensor_staging but need to be back-ported to master.

Test Plan: used by lazy_tensor_staging branch

Reviewed By: desertfire

Differential Revision: D32142032

fbshipit-source-id: f384ad7f1f054d122905afb1e82d7b71ee0b8ab2
